spec/osut_tests_spec.rb

```diff
@@ -1333,1134 +1129,2447 @@ module M
+    mini_north.setName("Mini north")
+    expect(mini_north.setSpace(core)).to be(true)
+    expect(mini_north.outsideBoundaryCondition).to eq("Outdoors")
+
+    vtx = OpenStudio::Point3dVector.new
+    vtx << mini_ceiling_vtx[3]
+    vtx << mini_floor_vtx[0]
+    vtx << mini_floor_vtx[1]
+    vtx << mini_ceiling_vtx[2]
+    mini_east = OpenStudio::Model::Surface.new(vtx, model)
+    mini_east.setName("Mini east")
+    expect(mini_east.setSpace(core)).to be(true)
+    expect(mini_east.outsideBoundaryCondition).to eq("Outdoors")
+
+    vtx = OpenStudio::Point3dVector.new
+    vtx << mini_ceiling_vtx[2]
+    vtx << mini_floor_vtx[1]
+    vtx << mini_floor_vtx[2]
+    vtx << mini_ceiling_vtx[1]
+    mini_south = OpenStudio::Model::Surface.new(vtx, model)
+    mini_south.setName("Mini south")
+    expect(mini_south.setSpace(core)).to be(true)
+    expect(mini_south.outsideBoundaryCondition).to eq("Outdoors")
+
+    vtx = OpenStudio::Point3dVector.new
+    vtx << mini_ceiling_vtx[1]
+    vtx << mini_floor_vtx[2]
+    vtx << mini_floor_vtx[3]
+    vtx << mini_ceiling_vtx[0]
+    mini_west = OpenStudio::Model::Surface.new(vtx, model)
+    mini_west.setName("Mini west")
+    expect(mini_west.setSpace(core)).to be(true)
+    expect(mini_west.outsideBoundaryCondition).to eq("Outdoors")
+
+    mini_floor.remove
+    mini_ceiling.remove
+    expect(mini_attic.setOutsideBoundaryCondition("Outdoors")).to be(true)
+
+    # Re-validating pre-tested areas + volumes.
+    expect(core.floorArea).to be_within(TOL).of(149.66 - 4) # -mini m2
+    core_volume = core.floorArea * core.ceilingHeight
+    expect(core_volume).to be_within(TOL).of(core.volume)
+    expect(attic.volume).to be_within(TOL).of(720.19)
+    expect(attic.floorArea).to be_within(TOL).of(567.98)
+
+    expect(mod1.poly(core_floor, true).empty?).to be(true)   # now concave
+    expect(mod1.poly(core_ceiling, true).empty?).to be(true) # now concave
+    expect(mod1.poly(attic_floor, true).empty?).to be(true)  # now concave
+    expect(mod1.logs.size).to eq(3)
+    expect(mod1.error?).to be(true)
+    mod1.logs.each {|l| expect(l[:message].include?("non-convex")).to be(true)}
+    expect(mod1.clean!).to eq(DBG)
+
+    file = File.join(__dir__, "files/osms/out/mini2.osm")
+    model.save(file, true)
+    # Simulation results E+ 22.2 (SutherlandHodgman, PolygonClipping)
+    # ... no E+ errors/warnings.
+    #
+    # mini2.osm A  FullInteriorAndExterior                 248 GJ 3.8 UMH cool
+    # mini2.osm B  FullExteriorWithReflections             249 GJ 3.7 UMH cool
+    # mini2.osm C  FullInteriorAndExteriorWithReflections  249 GJ 3.7 UMH cool
+
+    # --- --- --- --- --- --- --- --- --- --- --- --- --- --- --- --- --- --- #
+    # Mini as an attic "well":
+    #   - reintroduce mini floor @Z0 (ground-facing), as part of attic
+    #   - remove attic mini floor @Z3.05
+    #   - outdoor-facing courtyard walls become attic "well" walls.
+    #   - add interzone attic walls around well.
```

Comment:
Is your plan to algorithmically cut these zones out of the prototype models?  Or would you draw the attic geometry with skylights by hand, save as a separate OSM file, then swap all the geometry for the attic zone?  The second option is probably best for prototype models as long as the rest of the geometry is static.

Reply:
I guess my approach would be to make the core zone extend up into the attic (upside down mushroom) vs the attic extending down into the core zone.  That way the skylights are part of the core zone, and all the exterior loads (solar, conduction, infiltration) would hit the conditioned core zone directly (which I think is more realistic).

Reply:
**Q1**: Yes, it would be simpler to have a method adding skylights/wells in predetermined steps, specific to each _US DOE Prototype_. However, the goal here is to generalize its behaviour for arbitrary geometry (e.g. 3rd party models). What is being tested here are the lower-level methods. An eventual higher-level method would simply take in as argument the required _skylight-to-roof_ ratio (e.g. _SRR_ of 2% to 5%) for the entire model. If skylight wells/shafts are needed in the presence of plenums/attics, then they're autogenerated (in addition to the skylight sub-surfaces themselves). What is expected from 3rd party models are (i) valid model geometry and (ii) _spacetype_ allocation. Based on both, there is an intended prioritization of which spaces get skylights/wells. Something along these lines:
- first, spaces without side lighting (e.g. core spaces)
- preferably convex spaces, with significant floor area
- _spacetypes_ with a low probability of vertical obstructions or weird aspect ratios
- ... e.g. preferably classrooms, dining areas, lounges ... over kitchens, storage and corridors
- finally, spaces with high LPDs (greater potential savings)

There is a trial-and-error, iterative aspect to this:
- initially proportionately distribute potential skylight (& well) areas among spaces with similar (top) priority
- if conflicts arise (well doesn't _fit_ in a given _receiving_ ceiling), try a single redistribution of skylight areas elsewhere
- finally log a warning if still unable to meet _SRR_ targets

**Q2**: Yes, you're correct. A core zone would extend up the roof, as an upside-down mushroom (or a snub-nose beer bottle). These were simply tested variations on a theme with interpenetrating spaces/zones. Final tests will be done on the target cases you're describing.